pkg/reconciler/awscloudwatchsource/adapter.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright (c) 2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awscloudwatchsource
+
+import (
+	"context"
+	"encoding/json"
+
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	"knative.dev/pkg/logging"
+
+	"knative.dev/eventing/pkg/reconciler/source"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/kmeta"
+
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/reconciler/common"
+	"github.com/triggermesh/aws-event-sources/pkg/reconciler/common/resource"
+)
+
+const (
+	envRegion          = "AWS_REGION"
+	envQueries         = "QUERIES"
+	envPollingInterval = "POLLING_INTERVAL"
+)
+
+// adapterConfig contains properties used to configure the source's adapter.
+// These are automatically populated by envconfig.
+type adapterConfig struct {
+	// Container image
+	Image string `default:"gcr.io/triggermesh/awscloudwatchsource"`
+	// Configuration accessor for logging/metrics/tracing
+	configs source.ConfigAccessor
+	// Context for logging
+	context context.Context
```

Comment:
If it's just for logging, pass the logger directly. However, I'm not a big fan of having something that's not config in a struct called `adapterConfig`. For most cases you should be fine discarding serialization errors below, it is unlikely to happen because you're serializing something that's strongly typed.

If we need stronger validation, the proper way to report errors is by introducing a validation webhook IMO.

Reply:
So in other words, if relying on the k8s to validate based on the OpenAPI spec, then trust it.

Reply:
Even without validation at the API, when you reach this code, the K8S object's spec is a strongly typed Go struct that was already deserialized at some point (the client sent it as a YAML or JSON, eventually). Unless it contains functions, a struct can always be serialized to JSON without a custom serializer, so whatever the struct contains, this code will yield _some_ JSON.

If an error occurs due to some really odd error, I think the right thing to do is to panic for now, and add proper error handling to this function signature later (but then we need to update all sources).